### DISCUSSION TITLE
Same model giving different predictions in Python and C++

### DISCUSSION THREAD

**Discussion by sidrgupte** [dmlc/xgboost#12126] [Q&A] [answered]
I trained an XGB Regressor in Python and saved it as following:

```
trained_model = model.train_model(best_params)
json_path = os.path.join(manual_save_path, f"{target}.json")
booster = trained_model.get_booster()
booster.save_model(json_path)
```

Predictions: 

1. Just to be sure, I do the predictions in a separate file:
```
booster = xgb.Booster()
booster.load_model(json_path)
point = [-3.17492, 17.9437, 34.5295, 3.3425]
dtest_point = xgb.DMatrix([point])
pred = booster.predict(dtest_point)
print(f"Input vector: {point}\nPrediction: {pred[0]}")
```
I get the following:
>```
> Input vector: [-3.17492, 17.9437, 34.5295, 3.3425]
> Prediction: 1.0478190183639526
>```

2. Now, I do exactly the same in C++:
```
// test_xgb.cpp
#include <xgboost/c_api.h>
#include <iostream>
#include <vector>
#include <cmath> // for NAN

int main()
{
    BoosterHandle booster;
    if (XGBoosterCreate(nullptr, 0, &booster) != 0)
    {
        std::cerr << "Error creating booster!" << std::endl;
        return -1;
    }
    const char *model_path = "P.json";

    if (XGBoosterLoadModel(booster, model_path) != 0)
    {
        std::cerr << "Error loading model from " << model_path << std::endl;
        return -1;
    }

    std::vector<float> input = {-3.17492f, 17.9437f, 34.5295f, 3.3425f};
    std::cout << "Input vector: ";
    for (float f : input)
        std::cout << f << " ";
    std::cout << std::endl;
    DMatrixHandle dmat;

    if (XGDMatrixCreateFromMat(input.data(), 1, input.size(), NAN, &dmat) != 0)
    {
        std::cerr << "Error creating DMatrix!" << std::endl;
        return -1;
    }
    const char *json_config = R"({
    "type": 0,
    "training": false,
    "iteration_begin": 0,
    "iteration_end": 0,
    "strict_shape": true
})";

    const bst_ulong *out_shape;
    bst_ulong out_dim;
    const float *out_result;

    if (XGBoosterPredictFromDMatrix(
            booster,
            dmat,
            json_config,
            &out_shape,
            &out_dim,
            &out_result) != 0)
    {
        std::cerr << "Error predicting!" << std::endl;
        return -1;
    }
    bst_ulong out_len = 1;
    for (bst_ulong i = 0; i < out_dim; ++i)
    {
        out_len *= out_shape[i];
    }

    std::cout << "Prediction: ";
    for (bst_ulong i = 0; i < out_len; ++i)
    {
        std::cout << out_result[i] << " ";
    }
    std::cout << std::endl;

    XGDMatrixFree(dmat);
    XGBoosterFree(booster);

    return 0;
}
```

This gives me the following output:

> ```
> Input vector: -3.17492 17.9437 34.5295 3.3425 
> Prediction: 0.562144 
> ```

Python XGBoost version: 3.1.0
C++ XGBoost version: 3.1.0

I also tried saving the models as `.bin` file but did not work either!
Have I committed a silly mistake somewhere? Been stuck at this for the last 2 hours!

**Comment by trivialfis** [comment]
If the difference is significant, then the tree is taking different branches. I can't provide a definite answer without reproducing it myself. My guess is that the Python code uses f64 by default, and is converted to f32 inside XGBoost. During the text -> binary or during the f64 -> f32, the data is slightly changed due to floating-point parsing or conversion.

**Comment by trivialfis** [comment]
To illustrate, using the first number in your input:
```pycon
>>> from decimal import Decimal
>>> a = -3.17492
>>> Decimal(a)
Decimal('-3.1749200000000001864464138634502887725830078125')
```

**Reply by sidrgupte** [reply]
That is something I have already explored. Still, here is a snippet where I tried both - f32 and f64.
```
point64 = np.array([-3.17492, 17.9437, 34.5295, 3.3425], dtype=np.float64)
point32 = np.array([-3.17492, 17.9437, 34.5295, 3.3425], dtype=np.float32)

d64 = xgb.DMatrix(point64.reshape(1, -1))
d32 = xgb.DMatrix(point32.reshape(1, -1))

pred64 = e_p.predict(d64)[0]
pred32 = e_p.predict(d32)[0]

print(f"f64 input prediction: {pred64:.6f}")
print(f"f32 input prediction: {pred32:.6f}")
```

It gives:

> ```
> f64 input prediction: 1.047819
> f32 input prediction: 1.047819
>```

I am also attaching the model if you want to check it out. [P.json](https://github.com/user-attachments/files/26237331/P.json)

**Comment by trivialfis** [comment]
I couldn't reproduce the issue using the "P.json" you provided, tested with both 3.1 branch and the development branch.

```cpp
#include <xgboost/c_api.h>

#include <cmath>
#include <iostream>
#include <vector>

int main() {
    BoosterHandle booster;
    XGBoosterCreate(nullptr, 0, &booster);
    if (XGBoosterLoadModel(booster, "P.json") != 0) {
        std::cerr << "Error loading model: " << XGBGetLastError() << std::endl;
        return 1;
    }

    std::vector<float> input = {-3.17492f, 17.9437f, 34.5295f, 3.3425f};

    DMatrixHandle dmat;
    XGDMatrixCreateFromMat(input.data(), 1, input.size(), NAN, &dmat);

    const char *json_config = R"({
        "type": 0,
        "training": false,
        "iteration_begin": 0,
        "iteration_end": 0,
        "strict_shape": true
    })";

    const bst_ulong *out_shape;
    bst_ulong out_dim;
    const float *out_result;

    if (XGBoosterPredictFromDMatrix(booster, dmat, json_config,
                                    &out_shape, &out_dim, &out_result) != 0) {
        std::cerr << "Error predicting: " << XGBGetLastError() << std::endl;
        return 1;
    }

    std::cout << "Prediction: " << out_result[0] << std::endl;

    XGDMatrixFree(dmat);
    XGBoosterFree(booster);
    return 0;
}

```


```python
import numpy as np
import xgboost as xgb

booster = xgb.Booster()
booster.load_model("P.json")

point = [-3.17492, 17.9437, 34.5295, 3.3425]
dtest = xgb.DMatrix(np.array([point]))
pred = booster.predict(dtest, strict_shape=True)
print(f"Prediction: {pred[0]}")
```

**Reply by sidrgupte** [reply]
So you got the same result for both? What was it?

**Reply by trivialfis** [reply]
```
$ ls
P.json  repro  repro.cpp  repro.py
(xgboost_dev) [18:56:34] jiamingy@jiamingy-ord:~/ws/xgboost_dev/repro$ ./repro 
Prediction: 1.04782
(xgboost_dev) [18:56:36] jiamingy@jiamingy-ord:~/ws/xgboost_dev/repro$ python ./repro.py 
Prediction: [1.047819]
```

**Reply by sidrgupte** [reply]
I literally copy pasted your code and still getting the same 0.562144. Any idea what can it be?

**Reply by trivialfis** [reply]
C++ or both?

**Reply by sidrgupte** [reply]
I copied it for C++. Because for Python, we have the same answers.

**Reply by trivialfis** [reply]
Can you load the model in C++, save it again to a different file, and cross-compare the `base_score` parameter between the two JSON files?

**Reply by sidrgupte** [reply]
It remains the same i.e. ```5E-1```

**Reply by sidrgupte** [reply]
Let me rebuild my ```XGBoost```. I am not sure it would work but just to rule out.

**Reply by trivialfis** [reply]
> It remains the same i.e. 5E-1

I think you are using a different model.

The `P.json` has: `"base_score": "[9.856754E-1]",`

**Reply by sidrgupte** [reply]
Sorry, yes. This is being read correctly.
```
base_score: [9.856754E-1]
```

**Reply by trivialfis** [reply]
So, case closed?

**Reply by sidrgupte** [reply]
I have another ```base_score``` when I try to print the model config (pasting the particular dictionary from the model config)

```
"learner_model_param":
{"base_score":"5E1", 
"boost_from_average":"1",
"num_class":"0",
"num_feature":"4",
"num_target":"1"}
```

I feel the probably this is the ```base_score``` being used by my model.

**Reply by sidrgupte** [reply]
> So, case closed?

I am still getting the wrong prediction!

**Reply by trivialfis** [reply]
My focus on the base score is because the differences in the prediction output match:
```
>>> 9.856754E-1 - 0.5  # diff in base score
0.4856754
>>> 1.047819 - 0.562144  # diff in prediction
0.4856750000000001
>>> 
```
Aside from this, I don't have any suggestions. If you need help from me, I need a way to reproduce it precisely.

**Reply by sidrgupte** [reply]
I kind of understand the problem now. 
C++ is somehow finding the `base_score 0.5` in the dictionary `learner_model_param` and using this `base_score` to give the answer `0.562144`. But if I subtract this incorrect `base_score` from the answer I get  `0.0621439` and then when I add the true `base_score` i.e. ` 9.856754E-1`, I get `1.04781939` which is the correct answer!

And which begs the question, why is it using the wrong `base_score`?

I am using the same booster object to load the model: 

```
std::string model_path = "P.json";
BoosterHandle booster;
XGBoosterCreate(nullptr, 0, &booster);
XGBoosterLoadModel(booster, model_path.c_str());
```

The same file to find the `base_score` (This one prints the correct base_score)
```
std::ifstream model_file(model_path.c_str());
if (model_file.is_open())
{
    std::stringstream buffer;
    buffer << model_file.rdbuf();
    std::string json_text = buffer.str();
    model_file.close();
    if (pos != std::string::npos)
    {
        pos += 13; // skip past '"base_score":'
        std::size_t end = json_text.find(",", pos);
        std::string base_score_str = json_text.substr(pos, end - pos);
        std::cout << "base_score: " << base_score_str << std::endl;
    }
}
```
And the same booster to print the model config

```
 bst_ulong config_len;
  const char *config;
  if (XGBoosterSaveJsonConfig(booster, &config_len, &config) == 0)
  {
      std::cout << "\nModel config:\n"
                << config << std::endl;
  }

```

**Reply by trivialfis** [reply]
> C++ is somehow finding the base_score 0.5 in the dictionary learner_model_param and using this base_score to give the answer 0.562144

The base score is part of the learner model param, not entirely sure about the question here. Are you suggesting that the model param has base_score 9.856754E-1 but the C++ model uses 0.5 instead? I'm a bit confused at this point:
- You have different configurations as you posted the 0.5 base score before.
- I couldn't reproduce the error.

So, there's something I don't know about your pipeline.

**Reply by sidrgupte** [reply]
> The base score is part of the learner model param, not entirely sure about the question here. Are you suggesting that the model param has base_score 9.856754E-1 but the C++ model uses 0.5 instead? 

Yes! When I manually try to extract base_score from the json file, I get the correct base_score
```
// base score from json:
std::ifstream model_file(model_path.c_str());
if (model_file.is_open())
{
    std::stringstream buffer;
    buffer << model_file.rdbuf();
    std::string json_text = buffer.str();
    model_file.close();

    // look for "base_score" string in JSON
    std::size_t pos = json_text.find("\"base_score\":");
    if (pos != std::string::npos)
    {
        pos += 13; 
        std::size_t end = json_text.find(",", pos);
        std::string base_score_str = json_text.substr(pos, end - pos);
        std::cout << "base_score: " << base_score_str << std::endl;
    }
    else
    {
        std::cout << "base_score not found in model JSON." << std::endl;
    }
}
```

But when I ask XGB library to print me the config, it prints the wrong base_score.
```
// the model config:
bst_ulong config_len;
const char *config;
if (XGBoosterSaveJsonConfig(booster, &config_len, &config) == 0)
{
    std::cout << "\nModel config:\n"
              << config << std::endl;
}
```

**Reply by trivialfis** [reply]
I don't understand either. Just in case, could you please try the latest XGBoost as well?

**Reply by sidrgupte** [reply]
Well, I do kinda understand now. Because this is a known problem that was raised 3 years ago. Seems like it is still unresolved! Will try the latest version now.  

https://github.com/dmlc/xgboost/issues/9245

**Reply by trivialfis** [reply]
You referenced an issue with the evaluation metric. That's irrelevant to the base score

**Reply by sidrgupte** [reply]
I referenced an issue which is literally what I showed you above i.e. when I use `XGBoosterSaveJsonConfig` to print the `base_score` which is literally the first key in `learner_model_param` dictionary, it prints `0.5` which is not the correct value.

**Reply by trivialfis** [reply]
I used 3.1 to print the config as in your code, the model config looks right:
```
./repro 

Model config:
{"learner":{"generic_param":{"device":"cpu","fail_on_invalid_gpu_id":"0","n_jobs":"0","nthread":"0","random_state":"0","seed":"0","seed_per_iteration":"0","validate_parameters":"0"},"gradient_booster":{"gbtree_model_param":{"num_parallel_tree":"1","num_trees":"422"},"gbtree_train_param":{"process_type":"default","tree_method":"auto","updater":"grow_quantile_histmaker","updater_seq":"grow_quantile_histmaker"},"name":"gbtree","specified_updater":false,"tree_train_param":{"alpha":"0","cache_opt":"1","colsample_bylevel":"1","colsample_bynode":"1","colsample_bytree":"1","eta":"0.300000012","gamma":"0","grow_policy":"depthwise","interaction_constraints":"","lambda":"1","learning_rate":"0.300000012","max_bin":"256","max_cat_threshold":"64","max_cat_to_onehot":"4","max_delta_step":"0","max_depth":"6","max_leaves":"0","min_child_weight":"1","min_split_loss":"0","monotone_constraints":"()","refresh_leaf":"1","reg_alpha":"0","reg_lambda":"1","sampling_method":"uniform","sketch_ratio":"2","sparse_threshold":"0.20000000000000001","subsample":"1"},"updater":[{"hist_train_param":{"debug_synchronize":"0","extmem_single_page":"0","max_cached_hist_node":"18446744073709551615"},"name":"grow_quantile_histmaker"}]},"learner_model_param":{"base_score":"[9.856754E-1]","boost_from_average":"1","num_class":"0","num_feature":"4","num_target":"1"},"learner_train_param":{"booster":"gbtree","disable_default_eval_metric":"0","multi_strategy":"one_output_per_tree","objective":"reg:squarederror"},"metrics":[],"objective":{"name":"reg:squarederror","reg_loss_param":{"scale_pos_weight":"1"}}},"version":[3,1,3]}
Prediction: 1.04782
```

```cpp
/**
g++ -std=c++17 -o repro repro.cpp -I../xgboost/include -L../xgboost/lib -L$CONDA_PREFIX/lib -lxgboost -lgomp -Wl,-rpath,$(realpath ../xgboost/lib) -Wl,-rpath,$CONDA_PREFIX/lib
 */

#include <xgboost/c_api.h>

#include <cmath>
#include <iostream>
#include <vector>

int main() {
    BoosterHandle booster;
    XGBoosterCreate(nullptr, 0, &booster);
    if (XGBoosterLoadModel(booster, "P.json") != 0) {
        std::cerr << "Error loading model: " << XGBGetLastError() << std::endl;
        return 1;
    }

    // the model config:
    bst_ulong config_len;
    const char *config;
    if (XGBoosterSaveJsonConfig(booster, &config_len, &config) == 0)
    {
      std::cout << "\nModel config:\n"
                << config << std::endl;
    }

    std::vector<float> input = {-3.17492f, 17.9437f, 34.5295f, 3.3425f};

    DMatrixHandle dmat;
    XGDMatrixCreateFromMat(input.data(), 1, input.size(), NAN, &dmat);


    const char *json_config = R"({
        "type": 0,
        "training": false,
        "iteration_begin": 0,
        "iteration_end": 0,
        "strict_shape": true
    })";

    const bst_ulong *out_shape;
    bst_ulong out_dim;
    const float *out_result;

    if (XGBoosterPredictFromDMatrix(booster, dmat, json_config,
                                    &out_shape, &out_dim, &out_result) != 0) {
        std::cerr << "Error predicting: " << XGBGetLastError() << std::endl;
        return 1;
    }

    std::cout << "Prediction: " << out_result[0] << std::endl;

    XGDMatrixFree(dmat);
    XGBoosterFree(booster);
    return 0;
}

```

**Reply by sidrgupte** [reply, marked as answer]
Alright, It definitely had to do something with the older version. I downloaded the latest (`XGBoost version: 3.3.0`), did not touch a single thing in the old code and the problem is resolved! I am getting the correct predictions now. Thank you for your time though.

**Comment by trivialfis** [comment]
If you can't reproduce it using the scripts above, then we can narrow the issue down to something more specific:
- Are you copying the prediction output before the next call?
- Are you re-creating the DMatrix for new inputs?

**Reply by sidrgupte** [reply]
- I tried your code exactly as it is. Same old result.
- No. For now, I am just doing it for a single input.
